From 5400071e0896f7b92b4de2e48f17bd2aa1cde1fe Mon Sep 17 00:00:00 2001
From: Sekhar Nori <nsekhar@ti.com>
Date: Wed, 8 Aug 2012 16:29:42 +0530
Subject: [PATCH 02/23] davinci: da830/omap-l137 evm: add support for tsc2004
 controller

A new version of DA830 UI board is now available which adds
support for TSC2004 touchscreen controller.

This patch adds board specific hooks required to support the
touchscreen controller.

Signed-off-by: Sekhar Nori <nsekhar@ti.com>
---
 arch/arm/mach-davinci/board-da830-evm.c |   57 +++++++++++++++++++++++++++++++
 1 file changed, 57 insertions(+)

diff --git a/arch/arm/mach-davinci/board-da830-evm.c b/arch/arm/mach-davinci/board-da830-evm.c
index 21a096f..28bcd0f 100644
--- a/arch/arm/mach-davinci/board-da830-evm.c
+++ b/arch/arm/mach-davinci/board-da830-evm.c
@@ -18,6 +18,7 @@
 #include <linux/i2c.h>
 #include <linux/i2c/pcf857x.h>
 #include <linux/i2c/at24.h>
+#include <linux/i2c/tsc2004.h>
 #include <linux/mtd/mtd.h>
 #include <linux/mtd/partitions.h>
 #include <linux/spi/spi.h>
@@ -441,6 +442,53 @@ static struct pcf857x_platform_data __initdata da830_evm_ui_expander_info = {
 	.teardown	= da830_evm_ui_expander_teardown,
 };
 
+/*
+ * TSC 2004 Support
+ */
+#define TSC2004_GPIO_IRQ_PIN	GPIO_TO_PIN(2, 8)
+#define DA830_EVM_TSC2004_ADDRESS       0x49
+
+static int tsc2004_init_irq(void)
+{
+	int ret = 0;
+
+	ret = gpio_request(TSC2004_GPIO_IRQ_PIN, "tsc2004-irq");
+	if (ret < 0) {
+		pr_warning("%s: failed to TSC2004 IRQ GPIO: %d\n",
+								__func__, ret);
+		return ret;
+	}
+
+	ret = davinci_cfg_reg(DA830_GPIO2_8);
+	if (ret) {
+		pr_warning("%s: PinMux setup for GPIO %d failed: %d\n",
+			   __func__, TSC2004_GPIO_IRQ_PIN, ret);
+		return ret;
+	}
+
+	gpio_direction_input(TSC2004_GPIO_IRQ_PIN);
+
+	return ret;
+}
+
+static void tsc2004_exit_irq(void)
+{
+	gpio_free(TSC2004_GPIO_IRQ_PIN);
+}
+
+static int tsc2004_get_irq_level(void)
+{
+	return gpio_get_value(TSC2004_GPIO_IRQ_PIN) ? 0 : 1;
+}
+
+struct tsc2004_platform_data da830evm_tsc2004data = {
+	.model = 2004,
+	.x_plate_ohms = 180,
+	.get_pendown_state = tsc2004_get_irq_level,
+	.init_platform_hw = tsc2004_init_irq,
+	.exit_platform_hw = tsc2004_exit_irq,
+};
+
 static struct i2c_board_info __initdata da830_evm_i2c_devices[] = {
 	{
 		I2C_BOARD_INFO("24c256", 0x50),
@@ -455,6 +503,12 @@ static struct i2c_board_info __initdata da830_evm_i2c_devices[] = {
 	},
 };
 
+static struct i2c_board_info __initdata da830_evm_tsc2004_dev = {
+	I2C_BOARD_INFO("tsc2004", DA830_EVM_TSC2004_ADDRESS),
+	.platform_data  = &da830evm_tsc2004data,
+};
+
+
 static struct davinci_i2c_platform_data da830_evm_i2c_0_pdata = {
 	.bus_freq	= 100,	/* kHz */
 	.bus_delay	= 0,	/* usec */
@@ -592,6 +646,9 @@ static __init void da830_evm_init(void)
 	i2c_register_board_info(1, da830_evm_i2c_devices,
 			ARRAY_SIZE(da830_evm_i2c_devices));
 
+	da830_evm_tsc2004_dev.irq = gpio_to_irq(TSC2004_GPIO_IRQ_PIN),
+	i2c_register_board_info(1, &da830_evm_tsc2004_dev, 1);
+
 	ret = davinci_cfg_reg_list(da830_evm_mcasp1_pins);
 	if (ret)
 		pr_warning("da830_evm_init: mcasp1 mux setup failed: %d\n",
-- 
1.7.9.5

